doc/proposals/tls-utilities.md

```diff
@@ -0,0 +1,159 @@
+# TLS Utilities proposal
+
+## Background:
+
+A typical workflow of an operator that is written using the Operator-SDK involves watching a Custom Resource (CR), which specifies the desired state of the cluster, and making changes to the current state of the cluster to achieve the desired state.
+
+Each user specified CR represents a self contained system. Based on this, we can assume that any TLS assets generated are specific to each system which then is also specific to the corresponding CR.
+
+Based on the above assumption, we want to design a TLS utility pkg that helps generate TLS assets and ties them to a specific CR.
+
+## Goals:
+
+To provide TLS utilities for operator developers to create self signed Certificate Authority (CA) and the TLS encryption keys along with the signed certs. This will enable the operator to setup TLS assets for the application so that all the communication within it is secure.
+
+## Details:
+
+We will break down the TLS workflow into the following steps:
+
+1. Generate a Certificate Authority (CA): A CA is a centralized trusted third party that signs the certificates. We need to generate a CA key and CA Certificate to get started. If the users want to provide their own CA, we can add an option to do so. The generated CA certificate can be saved in a ConfigMap as follows:
+	
+	Inputs:
+
+	* CR-kind
+	* CR-name
+	* namespace
+
+	Outputs:
+
+        * A ConfigMap that contains the CA Cert:
+
+	```
+	kind: ConfigMap
+	apiVersion: v1
+	metadata:
+  	name: <crd-kind>-<crd-name>-ca
+  	namespace: <ns>
+	data:
+  	ca.crt: ...
+            ```
+        * A secret that contains the CA key:
+
+           ```
+ 	kind: Secret
+ 	apiVersion: v1
+ 	metadata:
+  	name: <cr-kind>-<cr-name>-ca
+  	namespace: <ns>
+ 	data:
+  	ca.key: ...
+	```
+
+2. Verify the CA certificate: The newly generated CA certificate in ConfigMap should be validated i.e. the digital signature should be checked, the expiry, activation dates and validity period should be checked, etc.
+
+3. Generate Server or Client Certificate: We assume that the components inside an application communicate through a service. Therefore, we generate a TLS private key and certificate for that service.
```

Comment:
If this service is an external service(using the external name) but I am using the operator to manage that, I would like the ability to generate a cert and set it for the external service. It appears from the implementation that this would not be possible and I think should be called out in this doc.

Reply:
@shawn-hurley could you explain what do you mean by `external service`? better if you can give me an concert example.

Reply:
Are you imaging that you have an external service of the endpoint `www.example.com` where the server is hosted somewhere else.
And you want to access this external service via the svc with `externalName`. e.g:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: my-service
  namespace: prod
spec:
  type: ExternalName
  externalName: www.example.com
```

Reply:
Yes that is what I was thinking about

Reply:
Who is managing the Cert from `www.example.com` since it doesn't live in the Kubernete cluster? I don't think tls util can create certs for `www.example.com`. Also, I am unsure how  do you manage cert for the external svc object.

I think the issue we are trying to solve is this. If a client accesses the  external svc `my-service.prod.svc.cluster.local`. It probably will get a cert for `www.example.com`returned with SAN field as `www.example.com`. The tls connection will fail because the client expects returned cert to have a SAN field  of `my-service.prod.svc.cluster.local` to match the connecting hostname.

Reply:
I agree with Fanmin.

I don't believe this fits in this proposal. This would need a separate proposal to fetch already generated custom certs and keys from k8s secrets.

Reply:
Sorry, maybe it was not clear what I was worried about. I think that if the type of service 'ExternalName` will not be able to be used to generate certs that will work then this should be noted in the proposal. 

Am I still misunderstanding what you guys are saying?

Reply:
Offline discussion:
The current TLS generation is designed to use the svc object as the mean to communicate components within a kubernetes cluster.
However, the user might want to generate cert for a component that's outside of the cluster. In that case, the generated cert needs to 
have custom SAN and that is not supported with current proposal. @shawn-hurley pointed out this use case. To resolve this, we can allow
user to add custom SAN as indicated in [certConfig](https://github.com/operator-framework/operator-sdk/blob/master/pkg/tlsutil/tls.go#L48).

Reply:
Also we should mention that if svc is a type of ExternalName, it won't work properly because because the SAN field for that svc needs additional custom fields from the user.

Reply:
Okay so to sum up the entire thread above:

@rithujohn191 Can you please add a note saying that GenerateCert() does not support the Service type `ExternalName`. The reason being that the FQDN of the Service object is used as the SAN for the certificate and not `spec.externalName`. Custom SANs will be supported later on through the CertConfig.